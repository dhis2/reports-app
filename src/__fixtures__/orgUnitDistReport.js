export const reportResponseData = {
    headers: [
        {
            name: 'ou',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
        {
            name: 'chp',
            column: 'CHP',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
        {
            name: 'chc',
            column: 'CHC',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
        {
            name: 'mchp',
            column: 'MCHP',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
        {
            name: 'clinic',
            column: 'Clinic',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
        {
            name: 'hospital',
            column: 'Hospital',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
    ],
    height: 16,
    rows: [
        ['Gbo', '', '1', '', '', ''],
        ['Bo', '18', '26', '58', '7', '4'],
        ['Tikonko', '1', '2', '3', '', ''],
        ['Niawa Lenga', '', '2', '1', '', ''],
        ['Bargbe', '', '1', '4', '', ''],
        ['Komboya', '', '1', '2', '', ''],
        ['Valunia', '2', '2', '3', '', ''],
        ['Lugbu', '2', '1', '6', '', ''],
        ['Selenga', '', '1', '1', '', ''],
        ['Wonde', '', '1', '2', '', ''],
        ['Bargbo', '1', '1', '5', '', ''],
        ['Kakua', '3', '7', '17', '7', '3'],
        ['Jaiama Bongor', '3', '1', '2', '', ''],
        ['Baoma', '3', '3', '7', '', ''],
        ['Bumpe Ngao', '3', '1', '4', '', '1'],
        ['Badjia', '', '1', '1', '', ''],
    ],
    width: 6,
    fileUrls: [
        {
            extension: 'xls',
            url:
                'http://localhost:8080/dhis/api/32/orgUnitAnalytics.xls?ou=O6uvpzGd5pu;YmmeuGbqOwR;JdhagCUEMbj;ARZ4y5i4reU;sxRd2XOzFbz;vWbkYPRmKyS;YuQRtpLP10I;I4jWcnFmgEC;daJPPxtIrQn;npWGUj37qDe;kU8vhUkAGaT;BGGmAwx33dj;dGheVylzol6;U6Kr7Gtpidn;KctpIIucige;zFDYIgyGmXG&ougs=J5jldMd8OHv&columns=J5jldMd8OHv',
        },
        {
            extension: 'pdf',
            url:
                'http://localhost:8080/dhis/api/32/orgUnitAnalytics.pdf?ou=O6uvpzGd5pu;YmmeuGbqOwR;JdhagCUEMbj;ARZ4y5i4reU;sxRd2XOzFbz;vWbkYPRmKyS;YuQRtpLP10I;I4jWcnFmgEC;daJPPxtIrQn;npWGUj37qDe;kU8vhUkAGaT;BGGmAwx33dj;dGheVylzol6;U6Kr7Gtpidn;KctpIIucige;zFDYIgyGmXG&ougs=J5jldMd8OHv&columns=J5jldMd8OHv',
        },
    ],
}
export const transformedTableData = {
    title: 'CHP - Bo',
    headers: [
        'Organisation Unit',
        'CHP',
        'CHC',
        'MCHP',
        'Clinic',
        'Hospital',
        'Total',
    ],
    rows: [
        ['Badjia', 0, 1, 1, 0, 0, 2],
        ['Baoma', 3, 3, 7, 0, 0, 13],
        ['Bargbe', 0, 1, 4, 0, 0, 5],
        ['Bargbo', 1, 1, 5, 0, 0, 7],
        ['Bumpe Ngao', 3, 1, 4, 0, 1, 9],
        ['Gbo', 0, 1, 0, 0, 0, 1],
        ['Jaiama Bongor', 3, 1, 2, 0, 0, 6],
        ['Kakua', 3, 7, 17, 7, 3, 37],
        ['Komboya', 0, 1, 2, 0, 0, 3],
        ['Lugbu', 2, 1, 6, 0, 0, 9],
        ['Niawa Lenga', 0, 2, 1, 0, 0, 3],
        ['Selenga', 0, 1, 1, 0, 0, 2],
        ['Tikonko', 1, 2, 3, 0, 0, 6],
        ['Valunia', 2, 2, 3, 0, 0, 7],
        ['Wonde', 0, 1, 2, 0, 0, 3],
        ['Bo', 18, 26, 58, 7, 4, 113],
    ],
}

export const chartResponseData = {
    headers: [
        {
            name: 'ou',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
        {
            name: 'chp',
            column: 'CHP',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
        {
            name: 'chc',
            column: 'CHC',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
        {
            name: 'mchp',
            column: 'MCHP',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
        {
            name: 'clinic',
            column: 'Clinic',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
        {
            name: 'hospital',
            column: 'Hospital',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
    ],
    height: 1,
    rows: [['Bo', '18', '26', '58', '7', '4']],
    width: 6,
    fileUrls: [
        {
            extension: 'xls',
            url:
                'http://localhost:8080/dhis/api/32/orgUnitAnalytics.xls?ou=O6uvpzGd5pu&ougs=J5jldMd8OHv&columns=J5jldMd8OHv',
        },
        {
            extension: 'pdf',
            url:
                'http://localhost:8080/dhis/api/32/orgUnitAnalytics.pdf?ou=O6uvpzGd5pu&ougs=J5jldMd8OHv&columns=J5jldMd8OHv',
        },
    ],
}

export const transformedChartData = {
    data: {
        labels: ['CHP', 'CHC', 'MCHP', 'Clinic', 'Hospital'],
        datasets: [
            {
                label: 'Count',
                backgroundColor: 'rgba(39, 102, 150, 0.6)',
                borderColor: 'rgb(39, 102, 150)',
                borderWidth: 1,
                hoverBackgroundColor: 'rgba(39, 102, 150, 0.8)',
                data: [18, 26, 58, 7, 4],
            },
        ],
    },
    options: {
        title: {
            display: true,
            text: 'CHP - Bo',
            fontSize: '16',
            fontFamily: "'Roboto', sans-serif",
        },
        legend: {
            position: 'bottom',
            fontFamily: "'Roboto', sans-serif",
        },
        animation: {
            duration: 180,
        },
    },
}
