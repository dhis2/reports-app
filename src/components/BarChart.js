import React from 'react'
import PropTypes from 'prop-types'
import { Bar } from 'react-chartjs-2'
import ReportLoader from './ReportLoader'

const BarChart = ({ content, isLoading }) => (
    <ReportLoader content={content} isLoading={isLoading}>
        <div className="chart">
            <Bar data={content.data} options={content.options} />
        </div>
    </ReportLoader>
)

BarChart.propTypes = {
    content: PropTypes.object.isRequired,
    isLoading: PropTypes.bool.isRequired,
}

export default BarChart
