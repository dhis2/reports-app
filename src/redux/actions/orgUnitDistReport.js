import { getOrgUnitDistReport } from '../../utils/api'
import {
    loadingReportDataStart,
    loadingReportDataSuccessWithFeedback,
    loadingReportDataErrorWithFeedback,
} from './reportData'

export const actionTypes = {
    SET_TABULAR_OUTPUT: 'SET_TABULAR_OUTPUT',
    SET_CHART_OUTPUT: 'SET_CHART_OUTPUT',
}
/**
 * @returns {Object}
 */
export const setOutputToChart = () => ({
    type: actionTypes.SET_CHART_OUTPUT,
})

/**
 * @returns {Object}
 */
export const setOutputToTabular = () => ({
    type: actionTypes.SET_TABULAR_OUTPUT,
})

/**
 * @returns {Function}
 */
export const loadChart = () => dispatch => {
    dispatch(setOutputToChart())
    dispatch(loadReport())
}

/**
 * @returns {Function}
 */
export const loadTable = () => dispatch => {
    dispatch(setOutputToTabular())
    dispatch(loadReport())
}

/**
 * @returns {Function}
 */
export const loadReport = () => (dispatch, getState) => {
    dispatch(loadingReportDataStart())

    const {
        organisationUnits,
        orgUnitGroupSets,
        orgUnitDistReport: { shouldShowChart },
    } = getState()
    const orgUnit = organisationUnits.selected
    const groupSetId = orgUnitGroupSets.selected

    return getOrgUnitDistReport(orgUnit, groupSetId)
        .then(response =>
            dispatch(
                loadingReportDataSuccessWithFeedback(response, shouldShowChart)
            )
        )
        .catch(error => dispatch(loadingReportDataErrorWithFeedback(error)))
}
