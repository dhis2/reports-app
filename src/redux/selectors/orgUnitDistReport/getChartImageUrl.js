import { getApi } from '../../../utils/api'

export const getChartImageUrl = state => {
    if (
        !state.organisationUnits.selected ||
        !state.orgUnitGroupSets.selected ||
        !state.orgUnitDistReport.displayImage
    ) {
        return ''
    }

    const { baseUrl } = getApi()
    const orgUnitId = state.organisationUnits.selected.id
    const orgUnitGroupSetsId = state.orgUnitGroupSets.selected
    const currentTime = new Date().getTime()

    const imageUrl = `${baseUrl}/organisationUnits/${orgUnitId}/distributionChart.png`
    const imageUrlParams = `groupSetId=${orgUnitGroupSetsId}&t=${currentTime}`

    return `${imageUrl}?${imageUrlParams}`
}
