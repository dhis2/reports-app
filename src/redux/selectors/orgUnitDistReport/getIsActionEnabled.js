export const getIsActionEnabled = state =>
    !!(
        state.organisationUnits.selected &&
        state.orgUnitGroupSets.selected &&
        !state.reportData.loading
    )
