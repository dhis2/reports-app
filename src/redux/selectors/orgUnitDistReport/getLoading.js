export const getLoading = state =>
    state.reportData.loading || state.orgUnitDistReport.chartImageLoading
