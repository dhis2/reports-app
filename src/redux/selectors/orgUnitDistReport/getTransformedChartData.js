import isEmpty from 'lodash.isempty'
import i18n from '@dhis2/d2-i18n'
import { getTitle } from '../../../utils/dataTransformHelpers'
import createDataTransformCache from '../../../utils/dataTransformCache'
import { FONT_FAMILY, BASE_COLOR } from '../../../config/style.config'

export const cache = createDataTransformCache()

/**
 * @param {Object} state - App state
 * @returns {Object} - Transformed table data that can be consumed by BarChart
 */
export default function getTransformedChartData(state) {
    const data = state.reportData.content

    if (isEmpty(data)) {
        return data
    }

    if (cache.hasValidCacheFor(data)) {
        return cache.getCachedResult()
    }

    const chartData = getChartData(
        getLabels(data.headers),
        getDataPoints(data.rows[0]),
        getTitle(state)
    )

    cache.setCachedResult(data, chartData)
    return chartData
}

function getLabels(headers) {
    return headers.reduce((acc, header, index) => {
        if (index > 0) {
            acc.push(header.column)
        }
        return acc
    }, [])
}

function getDataPoints(cells) {
    return cells.reduce((acc, cell, index) => {
        if (index > 0) {
            acc.push(parseInt(cell, 10))
        }
        return acc
    }, [])
}

function getChartData(labels, data, title) {
    return {
        data: {
            labels,
            datasets: [
                {
                    label: i18n.t('Count'),
                    backgroundColor: `rgba(${BASE_COLOR}, 0.6)`,
                    borderColor: `rgb(${BASE_COLOR})`,
                    borderWidth: 1,
                    hoverBackgroundColor: `rgba(${BASE_COLOR}, 0.8)`,
                    data,
                },
            ],
        },
        options: {
            title: {
                display: true,
                text: title,
                fontSize: '16',
                fontFamily: FONT_FAMILY,
            },
            legend: {
                position: 'bottom',
                fontFamily: FONT_FAMILY,
            },
            animation: {
                duration: 180,
            },
        },
    }
}
