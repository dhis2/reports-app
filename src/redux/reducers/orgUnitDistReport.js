import { actionTypes } from '../actions/orgUnitDistReport'

export const defaultState = {
    shouldShowChart: false,
}

export const orgUnitDistReport = (state = defaultState, { type } = {}) => {
    switch (type) {
        case actionTypes.SET_TABULAR_OUTPUT:
            return { shouldShowChart: false }

        case actionTypes.SET_CHART_OUTPUT:
            return { shouldShowChart: true }

        default:
            return state
    }
}
