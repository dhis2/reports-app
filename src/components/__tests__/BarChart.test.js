import React from 'react'
import { shallow } from 'enzyme'
import BarChart from '../BarChart'
import { transformedChartData } from '../../__fixtures__/orgUnitDistReport'

describe('<BarChart/>', () => {
    it('Should match the snapshot', () => {
        const wrapper = shallow(
            <BarChart content={transformedChartData} isLoading={false} />
        )
        expect(wrapper).toMatchSnapshot()
    })
})
