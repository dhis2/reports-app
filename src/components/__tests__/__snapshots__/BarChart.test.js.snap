// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BarChart/> Should match the snapshot 1`] = `
<ReportLoader
  content={
    Object {
      "data": Object {
        "datasets": Array [
          Object {
            "backgroundColor": "rgba(39, 102, 150, 0.6)",
            "borderColor": "rgb(39, 102, 150)",
            "borderWidth": 1,
            "data": Array [
              18,
              26,
              58,
              7,
              4,
            ],
            "hoverBackgroundColor": "rgba(39, 102, 150, 0.8)",
            "label": "Count",
          },
        ],
        "labels": Array [
          "CHP",
          "CHC",
          "MCHP",
          "Clinic",
          "Hospital",
        ],
      },
      "options": Object {
        "animation": Object {
          "duration": 180,
        },
        "legend": Object {
          "fontFamily": "'Roboto', sans-serif",
          "position": "bottom",
        },
        "title": Object {
          "display": true,
          "fontFamily": "'Roboto', sans-serif",
          "fontSize": "16",
          "text": "CHP - Bo",
        },
      },
    }
  }
  isLoading={false}
>
  <div
    className="chart"
  >
    <Bar
      data={
        Object {
          "datasets": Array [
            Object {
              "backgroundColor": "rgba(39, 102, 150, 0.6)",
              "borderColor": "rgb(39, 102, 150)",
              "borderWidth": 1,
              "data": Array [
                18,
                26,
                58,
                7,
                4,
              ],
              "hoverBackgroundColor": "rgba(39, 102, 150, 0.8)",
              "label": "Count",
            },
          ],
          "labels": Array [
            "CHP",
            "CHC",
            "MCHP",
            "Clinic",
            "Hospital",
          ],
        }
      }
      options={
        Object {
          "animation": Object {
            "duration": 180,
          },
          "legend": Object {
            "fontFamily": "'Roboto', sans-serif",
            "position": "bottom",
          },
          "title": Object {
            "display": true,
            "fontFamily": "'Roboto', sans-serif",
            "fontSize": "16",
            "text": "CHP - Bo",
          },
        }
      }
    />
  </div>
</ReportLoader>
`;
